File: 2904-shortest-and-lexicographically-smallest-beautiful-string/2904-shortest-and-lexicographically-smallest-beautiful-string.py
class Solution(object):
    def shortestBeautifulSubstring(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: str
        """
        ones = []

        # Store positions of all '1's
        for i, ch in enumerate(s):
            if ch == '1':
                ones.append(i)

        # Not enough 1s
        if len(ones) < k:
            return ""

        min_len = float('inf')
        answer = ""

        # Check every group of k consecutive 1s
        for i in range(len(ones) - k + 1):
            start = ones[i]
            end = ones[i + k - 1]

            substring = s[start:end + 1]
            length = len(substring)

            if length < min_len:
                min_len = length
                answer = substring

            elif length == min_len and substring < answer:
                answer = substring

        return answer

File: 2904-shortest-and-lexicographically-smallest-beautiful-string/2904-shortest-and-lexicographically-smallest-beautiful-string.cpp
class Solution {
public:
    string shortestBeautifulSubstring(string s, int k) {
        vector<int> ones;

        // Store positions of all '1's
        for (int i = 0; i < s.size(); i++) {
            if (s[i] == '1') {
                ones.push_back(i);
            }
        }

        // Not enough 1s
        if (ones.size() < k) {
            return "";
        }

        int minLen = INT_MAX;
        string answer = "";

        // Check every group of k consecutive 1s
        for (int i = 0; i <= ones.size() - k; i++) {
            int start = ones[i];
            int end = ones[i + k - 1];

            string substring = s.substr(start, end - start + 1);
            int length = substring.size();

            if (length < minLen) {
                minLen = length;
                answer = substring;
            }
            else if (length == minLen && substring < answer) {
                answer = substring;
            }
        }

        return answer;
    }
};